tests/src/runtimeApi/module/hipLaunchCooperativeKernel.cpp

```diff
@@ -0,0 +1,161 @@
+#include <iostream>
+#include <chrono>
+#include "test_common.h"
+
+using namespace std::chrono;
+
+const static uint BufferSizeInDwords = 448 * 1024 * 1024;
+
+__global__ void test_gws(uint* buf, uint bufSize, long* tmpBuf, long* result)
+{
+    extern __shared__ long tmp[];
+    uint offset = blockIdx.x * blockDim.x + threadIdx.x;
+    uint stride = gridDim.x  * blockDim.x;
+
+    long sum = 0;
+    for (uint i = offset; i < bufSize; i += stride) {
+        sum += buf[i];
+    }
+    tmp[threadIdx.x] = sum;
+
+    __syncthreads();
+
+    if (threadIdx.x == 0) {
+        sum = 0;
+        for (uint i = 0; i < blockDim.x; i++) {
+            sum += tmp[i];
+        }
+        tmpBuf[blockIdx.x] = sum;
+    }
+
+    // TODO replace this line with the sync function once it's available
+    __ockl_grid_sync();
+
+    if (offset == 0) {
+        for (uint i = 1; i < gridDim.x; ++i) {
+          sum += tmpBuf[i];
+       }
+       *result = sum;
+    }
+}
+
+int main() {
+  float *A, *B, *Ad, *Bd;
+  uint* dA;
+  long* dB;
+  long* dC;
+
+  uint32_t* init = new uint32_t[BufferSizeInDwords];
+  for (uint32_t i = 0; i < BufferSizeInDwords; ++i) {
+    init[i] = i;
+  }
+
+  hipDeviceProp_t deviceProp;
+
+  hipGetDeviceProperties(&deviceProp, 0);
+  if (!deviceProp.cooperativeLaunch) {
+    std::cout << "info: Device doesn't support cooperative launch! skipping the test!\n";
+    passed();
+    return 0;
+  }
+
+  std::cout << "info: running on bus 0x" << deviceProp.pciBusID << " " << deviceProp.name << "\n";
+
+  size_t SIZE = BufferSizeInDwords * sizeof(uint);
+
+  HIPCHECK(hipMalloc((void**)&dA, SIZE));
+  HIPCHECK(hipHostMalloc((void**)&dC, sizeof(long)));
+  HIPCHECK(hipMemcpy(dA, init, SIZE, hipMemcpyHostToDevice));
+
+  hipStream_t stream;
+  HIPCHECK(hipStreamCreate(&stream));
+
+  dim3 dimBlock = dim3(1);
+  dim3 dimGrid  = dim3(1);
+  int numBlocks = 0;
+  uint workgroups[2] = {32, 64};
+
+  system_clock::time_point start = system_clock::now();
+
+  for (uint i = 0; i < 2; ++i) {
+
+    dimBlock.x = workgroups[i];
+    // Calculate the device occupancy to know how many blocks can be run concurrently
+    hipOccupancyMaxActiveBlocksPerMultiprocessor(reinterpret_cast<uint32_t*>(&numBlocks),
+        test_gws, dimBlock.x * dimBlock.y * dimBlock.z, dimBlock.x * sizeof(long));
+
+    dimGrid.x = deviceProp.multiProcessorCount * std::min(numBlocks, 32);
+    HIPCHECK(hipMalloc((void**)&dB, dimGrid.x * sizeof(long)));
+ 
+    void *params[4];
+    params[0] = (void*)&dA;
+    params[1] = (void*)&BufferSizeInDwords;
+    params[2] = (void*)&dB;
+    params[3] = (void*)&dC;
+
+    std::cout << "Testing with grid size = " << dimGrid.x << " and block size = " << dimBlock.x << "\n";
+    HIPCHECK(hipLaunchCooperativeKernel(test_gws, dimGrid, dimBlock, params, dimBlock.x * sizeof(long), stream));
+
+    HIPCHECK(hipMemcpy(init, dC, sizeof(long), hipMemcpyDeviceToHost));
+
+    if (*dC != (((long)(BufferSizeInDwords) * (BufferSizeInDwords - 1)) / 2)) {
```

Comment:
The test should be fine in the sense that it can help to test the new API and catch some potential regressions. There are rooms for improving the test to catch some corner cases and make it stronger. These improvements and adding more tests will be done in follow up PRs.